tt_metal/impl/device/device_mesh.cpp

```diff
@@ -119,13 +160,40 @@ std::vector<Device*> DeviceMesh::get_devices_on_column(int col_idx) const {
 
 const DeviceIds DeviceMesh::get_device_ids() const
 {
+    // Return the logical device ids
     DeviceIds device_ids;
     for (const auto& [device_id, device] : mesh_devices) {
+        // device_id is not always equal to device->id()
         device_ids.push_back(device_id);
     }
     return device_ids;
 }
 
+std::vector<int> DeviceMesh::get_ring_device_indices(
```

Comment:
@cfjchu let's iterate on what this should look like. My first pass at it is to take in a vector of Devices and return the indices that will get you a ring. I've thought about maybe modifying the vector of Devices in place and returning that back?

Reply:
So this is great for when we construct the DeviceMesh. It frees the user from having to know the device-id ordering when constructing the mesh. So what I would ideally like is a way to say: 

`device_mesh = DeviceMesh(MeshShape(2,4), topology=Topology::RING)`

internally in the constructor, open all 8 devices and then reorder the devices in a ring and have the 2d logical view map to this ring.

Reply:
So you're thinking of not exposing `get_device_ring_idx` at all? Then the api can be private to DeviceMesh

Reply:
So we need your function there to dynamically figure out device-ids in a ring and remove all the hardcoding on `device_ids = [0, 4, 5, 1, 2, 6, 7, 3]` for T3K. 

Those device ids will then get mapped onto a 2D logical mesh. Then, if we want to query to get a view of a 2x2 ring on the 2x4 DeviceMesh, then we should have additional APIs on the `tt_metal/impl/device/device_mesh_view.hpp` class to do something like:

```
view = device_mesh->get_view();
sub_device_ring = view->get_devices(start=Coordinate{.row=0, .col=0}, MeshShape={2,2}, TraversalOrder::Ring);

// dispatch ttnn.all_gather(..) on the subdevice_ring
```

Reply:
Are we able to remove the T3K hardcoding of device-ids using your function now?

Reply:
Makes sense. I'll touch up the PR to fix the hardcoded T3k values first.